### ISSUE TITLE
Unexpected behaviour of Retry feature using some RDBMSs versions

### ISSUE THREAD

**Issue by rubens21** [sequelize/sequelize#8352]
## What are you doing?
The Retry feature fails to identify a rolled back transaction caused by a deadlock (or there is a lack of documentation).

The default behaviour of some MySQL and MariaDb versions is rolling back a transaction if a dead lock is detected (https://dev.mysql.com/doc/refman/5.7/en/innodb-deadlock-detection.html). 
Although it may be changed by configuration option (MySQL 
https://dev.mysql.com/doc/refman/5.7/en/innodb-parameters.html#sysvar_innodb_deadlock_detect, and MariaDb >= 10.2.26 https://mariadb.com/kb/en/library/xtradbinnodb-server-system-variables/#innodb_deadlock_detect), the default action is rolling back the whole transaction, not only the last query.

Due this behaviour, the Retry option may hide an error because the Transaction Promise chain is not broken even after the transaction in the database has been rolled back.

In summary, the Sequelize Transaction promise chain may finish successfully even when all data has been lost.

```js
const Promise = require('bluebird');
const Sequelize = require('sequelize');

let sequelize = new Sequelize('maydb', 'myuser', 'a-secure-password',
  {
    host: 'localhost',
    dialect: 'mysql',
    retry: {
      match: /Deadlock/i,
      max: 5
    },
    logging: console.log
  });

/*
    CREATE TABLE foo (col1 TEXT);
    CREATE TABLE tableA (col1 INT);
    INSERT INTO tableA VALUES (0);
    CREATE TABLE tableB (col1 INT);
    INSERT INTO  tableB VALUES (0);
     */

let update = (t, table, value, delay) => {
  return new Promise((resolve, reject) => {
    setTimeout(() => {
      console.log(`${table} -> ${value};`);
      sequelize
      .query(`UPDATE ${table} SET col1 = ${value};`, {transaction: t})
      .then(resolve)
      .catch(reject);
    }, delay);
  });
};

let process1 = () => {
  return sequelize.transaction((t1) => {
    return writeInFoo("Process1 has started", t1).then(()=> {
      return update(t1, 'tableA', 1, 0);//first update: timer to keep the right order to force a dead lock
    }).then(() => {
      return writeInFoo("Process1 changed tableA", t1);
    }).then(() => {
      return update(t1, 'tableB', 1, 10);//fourth update: timer to keep the right order to force a dead lock
    }).then(() => {
      return writeInFoo("Process1 changed tableB", t1);
    });
  }).then(() => {
    return true;
  }).catch((err) => {
    console.error(err);
    return false;
  });
};

let process2 = () => {
  return sequelize.transaction((t2) => {
    return writeInFoo("Process2 has started", t2).then(()=> {
      return update(t2, 'tableB', 2, 5);//second update: timer to keep the right order to force a dead lock
    }).then(() => {
      return writeInFoo("Process2 changed tableB", t2);
    }).then(() => {
      return update(t2, 'tableA', 2, 0);//third to keep the right order to force a dead lock
    }).then(() => {
      return writeInFoo("Process2 changed tableA", t2);
    });
  }).then(() => {
    return true;
  }).catch((err) => {
    console.error(err);
    return false;
  });
};

let writeInFoo = (message, t) => {
  return sequelize.query(`INSERT INTO foo VALUES (:msg);`, {
    replacements: {msg: message},
    type: sequelize.QueryTypes.INSERT,
    transaction: t
  });
};

Promise.all([process1(), process2()]).then((r) => {
  console.log(r);
  process.exit()
});

```

## What do you expect to happen?
Both process should be atomic.

1. The table `foo` should have 3 or 6 entries. (all success or only one success)
2. The tables `tableA` and `tableB` should have all entries with the same value (all 1 or all 2).
3. If one of the process has been failed, the final result should have one FALSE value.
4. The MySQL/MariaDb general log should not have a COMMIT query in a thread which lost the transaction.

## What is actually happening?
1. The table `foo` has 4 entries (3 from the succeeded process, and the last one of the failed process).
![image](https://user-images.githubusercontent.com/447656/30705430-70074a3e-9ec3-11e7-9ac5-ce71d946b7de.png)

2. The tables `tableA` and `tableB` has different values because the last update was made by the failed process.
3. Both process returns TRUE

Script output:
![image](https://user-images.githubusercontent.com/447656/30705479-910f8e9e-9ec3-11e7-83c3-5052b08788b0.png)

4. Both treads has all query executed, even after one of them has been rolled back.
Note that there are two `UPDATE tableB SET col1 = 1` been executed! The second one is the second try done by Sequelize, and the transaction of the thread 3231 has already been rolled back at this moment.

![image](https://user-images.githubusercontent.com/447656/30705388-4e2f0fb4-9ec3-11e7-8bf8-a5ecc3da456a.png)
 

__Dialect:__ mysql
__Dialect version:__ 2
__Database version:__ 10.0.31-MariaDB-1~trusty-wsrep mariadb.org binary distribution, wsrep_25.19.rc3fc46e
__Sequelize version:__ 4.9.0
__Tested with master branch:__ Yes

**Comment by depoulo** [comment]
I think you might have misunderstood the documentation.
You have to wrap the whole transaction into `retry-as-promised`, **not** configure Sequelize's `retry` option globally to catch deadlocks (and nothing else^^).

See the discussion at this thread here: https://github.com/sequelize/sequelize/issues/1478#issuecomment-132099311

Example:

```js
import Sequelize from 'sequelize';
import Bluebird from 'bluebird';
import retry from 'retry-as-promised';

function runQueriesInTransactionSequence(queries) {
    return (sequelize, options) => {
        return retry(() => {
            return sequelize.transaction(options, transaction => {
                return Bluebird.each(queries, statement => {
                    return sequelize.query(statement, Object.assign({}, options, {transaction}));
                });
            });
        }, {
            max: 10,
            timeout: 10000,
            match: [
                Sequelize.ConnectionError,
                Sequelize.ConnectionRefusedError,
                Sequelize.ConnectionTimedOutError,
                Sequelize.OptimisticLockError,
                Sequelize.TimeoutError,
                /Deadlock/i
            ]
        });
    };
}
```

**Comment by depoulo** [comment]
@ Sequelize maintainers: I don't know if what I wrote is correct, but you might acknowledge that there is a lot of confusion regarding this topic, so maybe the information that (and why) you have to manually retry transactions, together with a good example (better than the one by me above) should go into the documentation.

**Comment by rubens21** [comment]
@depoulo Thank you for your explanation.
As I had mentioned in the original post, one of the problem may be "lack of documentation". I did not found any documentation explaining how it works. Sorry if it was my fault.

Anyway, I do believe that Sequelize should 
(a) Detect that the transaction has been rolled back 
(b) or having a flag to force developers be aware of the risks of the transaction do not be atomic
(c) or having a very visible warning on the documentation talking about this risk;

Even if this behaviour was not a bug, Sequelize should not allow that weird behaviour be transparent unless an option like (b) is available.

Notes about your comments:

> I think you might have misunderstood the documentation.

Please, could you help me to find more documentation about it? I just found the option description

> You have to wrap the whole transaction into retry-as-promised, not configure Sequelize's retry option globally to catch deadlocks (and nothing else^^).

If you are right, Sequelize should not have this option since we cannot use it.

Thank you

**Comment by depoulo** [comment]
Hey again,
the only documentation I could find, and this is when I started wrapping all my transactions into retry-as-promised, is the discussion I linked above under "#1478 (comment)".

Hope that helps.

**Comment by rubens21** [comment]
I also had found it: http://docs.sequelizejs.com/class/lib/sequelize.js~Sequelize.html

On both places it is not clear enough about its risks. 
I think it is not exactly a bug, we just need to know that Sequelize will keep going after the transaction has been rolled back after a dead lock.

**Comment by schmod** [comment]
This is still an issue in Sequelize 5, and can actually be an issue even when `retry` is disabled.

If I'm running a few operations in parallel, or am manually handling the transaction, it's possible that I will issue further queries against the transaction (which Sequelize still believes to be active) after the deadlock (and implicit rollback) happens.  

This can result in `INSERT` and `UPDATE` statements being inadvertently committed to the database. 

---

AFAIK, the only way to correctly handle a MySQL deadlock inside of a transaction is for Sequelize to immediately mark the transaction as complete, and to throw an error if we attempt to issue further queries against the (now non-existent) transaction.

This won't fix the retry behavior (which seems like it'll be effectively impossible to implement at the ORM level), but it will at least cause these retries to throw errors in scenarios where they are unsafe.

**Comment by sushantdhiman** [comment]
Fixed by https://github.com/sequelize/sequelize/pull/11074
